### PR TITLE
quick access: improve hover effect

### PR DESCRIPTION
**Pull request by elstoc**
As suggested in https://github.com/darktable-org/darktable/pull/8247#issuecomment-782109070

### CODE REVIEW

**Comment by elstoc** [comment]
@Nilvus - the buttons need some padding but it's hard to easily do this while consistently retaining the size of the icons. Please can you take a look at what I've done and suggest any improvements. I think I've probably removed some of the compactness (because the headers are taller)

File: data/themes/darktable.css [hunk 1]
```diff
@@ -1539,8 +1539,9 @@ margin: 0 2px;
 {
   min-height: 0.95em;
   min-width: 0.95em;
-  margin: 0 0.75em;
+  margin: 0 0.45em;
   border: none;
+  padding: 0.3em;
```
Comment (Nilvus) [inline data/themes/darktable.css:1544]: `padding: 0.2em 0.3em 0.3em 0.2em;`

File: data/themes/darktable.css [hunk 2]
```diff
@@ -1539,8 +1539,9 @@ margin: 0 2px;
 {
   min-height: 0.95em;
   min-width: 0.95em;
-  margin: 0 0.75em;
+  margin: 0 0.45em;
```
Comment (Nilvus) [inline data/themes/darktable.css:1542]: It's now me who find toggle button too small. 0.5em would be better.
Set `margin: 0 0.5em;`

**Review by Nilvus — CHANGES_REQUESTED** [review]
It's a weird thing to align hover effect with padding. Most of the times, You need to set them differently, with top=left and bottom=right and those 2 twins a little different. See my proposal to understand.

File: data/themes/darktable.css (hunk 1 above)
Comment (elstoc) [reply data/themes/darktable.css:1544]: fine with this - it does indeed make the borders more even.

File: data/themes/darktable.css (hunk 2 above)
Comment (elstoc) [reply data/themes/darktable.css:1542]: This makes my buttons smaller than in master when I use a smaller font. I mostly got to 0.45 by trial and error comparison with master.

Comment (elstoc) [reply data/themes/darktable.css:1542]: I think the extra padding is taking space from the height. If I use 0.5em padding and resize the icons to min 1.15em it looks the same as master.

**Comment by elstoc** [comment]
I've pushed a change that seems to replicate the layout and sizes in current master, but with the extra hover effect.

**Comment by Nilvus** [comment]
> I've pushed a change that seems to replicate the layout and sizes in current master, but with the improved hover effect.

We were agree on my CSS PR for 0.95em and that 1.15em was too high. Please revert it. Setting 0.5em instead of 0.45em was sufficient.

File: data/themes/darktable.css [hunk 3]
```diff
@@ -1537,10 +1537,11 @@ margin: 0 2px;
 #basics-box-labels #module-enable-button,
 #basics-link
 {
-  min-height: 0.95em;
-  min-width: 0.95em;
-  margin: 0 0.75em;
+  min-height: 1.15em;
+  min-width: 1.15em;
```
Comment (Nilvus) [inline data/themes/darktable.css:1541]: To revert back.

**Review by Nilvus — CHANGES_REQUESTED** [review]

File: data/themes/darktable.css (hunk 3 above)
Comment (elstoc) [reply data/themes/darktable.css:1541]: Just do a comparison with master. The sizes of the actual icons are identical between my PR (latest commit) and master. I'm telling you that the padding added by this change reduces the size of the icons. All I'm doing is compensating for that reduction.

Comment (elstoc) [reply data/themes/darktable.css:1541]: The only think I can conclude here is that there are some boundaries in em where things like icons change by whole pixel values. With my change, at font sizes 11 and 12 I agree the icons look a little bigger. With smaller font sizes (10 or below), the icons look the same to me. I don't know how to come up with a good compromise here.

Comment (Nilvus) [reply data/themes/darktable.css:1541]: That's not what I have on my side. See below (from left to right: master, your PR and your PR with 0.95em)

Comment (elstoc) [reply data/themes/darktable.css:1541]: 1.05em is virtually indistinguishable from master at 9-12em so seems a good compromise

Comment (elstoc) [reply data/themes/darktable.css:1541]: @Nilvus what font size do you use

Comment (elstoc) [reply data/themes/darktable.css:1541]: here's what I see with font size 8.4

master, followed by my PR with 0.95 
Screenshot_2021-02-19_20-19-54Screenshot_2021-02-19_20-20-20

Comment (Nilvus) [reply data/themes/darktable.css:1541]: Forget all that. I really don't understand why I had UI shown in screenshots above, so issue I pointed here, but after removing completely the folder where I had the clone of your repository, I have now correct UI and can confirm that 1.15em is good. I suspect something not updated correctly in CSS file in source code. It's now ok. Sorry for the noise!

**Review by Nilvus — APPROVED** [review]

**Comment by Nilvus** [comment]
@TurboGit: So all good for me here. You could merge.

**Comment by TurboGit** [comment]
Merging then, thanks !

File: data/themes/darktable.css (hunk 3 above)
Comment (elstoc) [reply data/themes/darktable.css:1541]: That's ok. There's still some odd interaction here that I don't understand though, and I do notice a difference on larger fonts. The whole UI should scale evenly if everything is in ems. But changing the left/right margin shouldn't change the size of the icon when there's plenty of free space on either side of it. I'm assuming either there's something hard-coded in px somewhere or Gtk just isn't respecting the box model properly.

**Comment by dterrahe** [comment]
Apologies if this is an annoying place to start a discussion, but I've been wondering for a while if it were possible to refactor the CSS code. It seems there are a lot of hard-coded specific small adjustments for individual widgets, for example buttons, when presumably what we want to achieve is consistent sizing everywhere.

At first glance, it looks like we mostly have two distinct button sizes;
- for ui buttons (top row, bottom row, module group buttons)
- module buttons (lib & iop; headers (on/off/instance/(p)reset), color pickers, masking, drawing)
- some exceptions, for example overlay stars etc

Some of these have slightly differing sizes within a category, but should they have?

I don't know CSS enough to see where this might fall down, but could their sizes not be set at the top and then adjusted down at the panel level and then automatically cascade down? Or alternatively have a set of master definitions that get reused consistently? That way it would be possible to add a theme with bigger buttons, for example, without having to literally redefine (and keep up-to-date) _everything_?

Another side-effects of the current implementation is that when adding a new widget it rarely defaults to some reasonable look, because none have been set; all the others have been individually tweaked, so now each new one needs a tweak _too_. Consistency is hard to maintain that way and I think there is some room for improvement there.

Its fine and to be expected that we ended up where we are, but after a system has been used for a while and it has become clearer than in the beginning what it should achieve, there comes a time to take a step back and see if that goal can be achieved in a cleaner way, rather than continuing down the same path and calcifying the existing approach. I wouldn't be surprised if a more compact CSS could also lead to some speed/memory improvements (although very much secondary).

**Comment by elstoc** [comment]
I'm sure you're right and some refactoring can be done here. I suspect it will probably need some Gtk changes as well - I think we probably have some explicitly named widgets that can be styled by targeting those widgets more generically. I've been considering allowing more compact layouts out-of-the box as well - I currently have my own CSS tweaks to reduce the size of the module header, for example. We could do this by separating the theme in two: color scheme & layout.

It's worth some discussion anyway - perhaps you could raise a new FR with a summary.

**Comment by dterrahe** [comment]
I'm _not at all_ sure I'm right! ;-)

On my todo list there's the item "become a gtk css specialist so I can start saying sensible things about it". But unfortunately it's way down. An FR is a very public place to say silly things about stuff you know little about; much safer in the comments of a closed PR.

**Comment by Nilvus** [comment]
I agree that some parts would need to be sure to have same size but already many things are already in that case. What's difficult is that Gtk code is not always consistent and so CSS tweaks are needed. The whole UI is quite consistent by now but of course this has been improved. In last spring, I have done a lot of work to do what you suggest @dterrahe and so reduced/optimize CSS code (with lot of comments to better read it). But we have seen since many update/adds in the UI. So the work should be done again for all things added since.

@elstoc: I'm interesting in your CSS tweaks to test them. Could you share?
About separate theme (color scheme & layout), I disagree. This would overload theme selector and if doing that in CSS file has its limits as CSS order is important for many things to be correctly applied. This will be worse and harder to do and maintain.

I always try to optimize CSS as more as I can. But my limit is I can't update Gtk code, except really few things. I'm on if you want to work on it and help. Will check ASAP what's CSS can be grouped on all recent updated things.

@dterrahe: I can do the CSS work, I'm sure your skills could be better move on many parts and/or help on what you suggest.

To sum, work again on CSS code to optimize it is already in my todo-list. I've just find a way to regroup basics settings with same module one. Just remains one thing to fix by trying to not add CSS code. It's a first step.

**Comment by elstoc** [comment]
> I'm interesting in your CSS tweaks to test them.

Apart from hacking the lighttable to always show stars and color labels on top of the image, my main condensing tweak is `#module-header {padding: 0.23em;}`

**Comment by Nilvus** [comment]
Two other things to answer here:

Some things needs to have slightly different sizes, like for example here, widget module name between module header name, as it's not same thing. But many things are already different.

@dterrahe: you say that when adding a widget we rarely have some reasonable look. I disagree with that. When adding some tweaks, since my big CSS work last year, it's always minor things that I work on and I always see that all the main things and about 80 to 90% of the work is applied correctly without doing any CSS tweaks.

But with your comment and what I've just checked about what @elstoc fix in that PR is that in fact, main of CSS code here was already and not use, just because on start GTK Code was not good for those buttons. If you just remove, CSS code now that Gtk code have been removed, you will see that just margin are bad. But if #basics-label-box was set with #module-header, you will see that margin is good, except top margin of all #basics box.

See, if I remove CSS of that PR and set #basics-box-label with #module-header margin:



All that make me suggest on thing from all that. Maybe we would need an article about UI design and how to deal with that, beginning by using same CSS tags for all same part and be sure, before adding any new tag. A good CSS code begin with that and so adapted Gtk code. So a guideline is needed I think.

**Comment by Nilvus** [comment]
> #module-header {padding: 0.23em;}

Thanks for the share. You seems to like tight UI.

**Comment by elstoc** [comment]
> You seems to like tight UI

I don't like scrolling and I like as much space as possible for my image.

**Pull request merged**